pkg/genericapiserver/enablement_config.go

```diff
@@ -0,0 +1,157 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package genericapiserver
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+type APIResourceEnablement interface {
+	ResourceEnabled(resource unversioned.GroupVersionResource) bool
+	VersionEnabled(version unversioned.GroupVersion) bool
+}
+
+// Specifies the overrides for various API group versions.
+// This can be used to enable/disable entire group versions or specific resources.
+type APIGroupVersionEnablementConfig struct {
+	// Whether to enable or disable this entire group version.
+	Enable bool
+
+	// DisabledResources lists the resources that are specifically disabled for a group/version
+	// DisabledResources trumps EnabledResources
+	DisabledResources sets.String
+
+	// EnabledResources lists the resources that should be enabled by default.  This is a little
+	// unusual, but we need it for compatibility with old code for now.  An empty set means
+	// enable all.
+	EnabledResources sets.String
+}
+
+// AnyVersion is used during resource enablement checks to see if any version of the resource is enabled
+const AnyVersion = "."
+
+type APIResourceEnablementConfig struct {
+	GroupVersionDisablers map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig
+}
+
+func NewAPIResourceEnablementConfig() *APIResourceEnablementConfig {
+	return &APIResourceEnablementConfig{GroupVersionDisablers: map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig{}}
+}
+
+func NewAPIGroupVersionEnablementConfig() *APIGroupVersionEnablementConfig {
+	return &APIGroupVersionEnablementConfig{Enable: true, DisabledResources: sets.String{}, EnabledResources: sets.String{}}
+}
+
+func (o *APIResourceEnablementConfig) DisableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = false
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = true
+	}
+}
+
+func (o *APIResourceEnablementConfig) DisableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Delete(resource.Resource)
```

Comment:
this could make `EnabledResources` be empty and therefore enable all other unspecified resources, right?